### ISSUE TITLE
Unable to parse Plugin Plists

### ISSUE THREAD

**Issue by Ghawken** [FlyingDiver/Indigo-Masquerade#12]
Hi

Version:2022.0.1

Not sure whether this is user error or not - just fiddling to see best way to get run plugin action from within another plugin and seeing this.  

Thought would post before - can copy and paste easily...

```
Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Masquerade.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/GhostXML.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Virtual Devices.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/SunRise.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Relay and Dimmer.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Hue Lights.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Thermostat.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Email+.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Sprinkler.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Cynical SecuritySpy.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/EPS Device Extensions.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Occupatum.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/HomeKitLink-Siri.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/iMessage.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Masquerade.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/GhostXML.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Virtual Devices.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/SunRise.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Relay and Dimmer.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Hue Lights.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Thermostat.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Email+.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Sprinkler.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Cynical SecuritySpy.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/EPS Device Extensions.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Occupatum.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/HomeKitLink-Siri.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/iMessage.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Masquerade.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/GhostXML.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Virtual Devices.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/SunRise.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Relay and Dimmer.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Hue Lights.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Thermostat.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Email+.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Sprinkler.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Cynical SecuritySpy.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/EPS Device Extensions.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Occupatum.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/HomeKitLink-Siri.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/iMessage.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Masquerade.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/GhostXML.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Virtual Devices.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/SunRise.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Relay and Dimmer.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Hue Lights.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Thermostat.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Email+.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Example Device - Sprinkler.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Cynical SecuritySpy.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/EPS Device Extensions.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/Occupatum.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/HomeKitLink-Siri.indigoPlugin/Contents/Info.plist
   Masquerade Warning              getActionFieldList: Unable to parse plist, skipping: /Library/Application Support/Perceptive Automation/Indigo 2022.1/Plugins/iMessage.indigoPlugin/Contents/Info.plist
```

Adding exception catching to code:

Gives this reason:
` err module 'plistlib' has no attribute 'readPlist'`

**Comment by FlyingDiver** [comment]
Unable to replicate.  That sure looks like you were running the Py2 plugin under Py3 (somehow).  plistlib.readPlist() was in the Py2 plugin, and was replaced in the Py3 version.

**Comment by Ghawken** [comment]
Hmmm.  That certainly would be a more interesting bug as to how that happened/was possible

```
Stopped plugin Masquerade
   Installing plugin Masquerade v2022.0.1 (previous version 2022.0.1 moved to trash)
   Loading plugin "Masquerade 2022.0.1" using API v3.0
   Starting plugin "Masquerade 2022.0.1" (pid 84671)
```

Tried PluginStore version and latest GITHUB version..
Will stop, delete plugin folder info and try again..

**Comment by Ghawken** [comment]
Updated to latest - supportURL and problem is gone.

Looking at prior python 3 plugin store version 2022.0.1  version — plistlib.readPlist() seems to be there….

**Comment by Ghawken** [comment]
Seems to be here on github version as well?

```
def getActionList(self, filter="", valuesDict=None, typeId="", targetId=0):
        retList = []
        indigoInstallPath = indigo.server.getInstallFolderPath()
        pluginsList = os.listdir(indigoInstallPath + '/Plugins')
        for plugin in pluginsList:
            if (plugin.lower().endswith('.indigoplugin')) and (not plugin[0:1] == '.'):
                path = f"{indigoInstallPath}/Plugins/{plugin}/Contents/Info.plist"
                try:
                    pl = plistlib.readPlist(path)
                except(Exception, ):
                    self.logger.warning(f"getActionList: Unable to parse plist, skipping: {path}")
                else:
                    bundleId = pl["CFBundleIdentifier"]
                    if bundleId == valuesDict.get("devicePlugin", None):
                        self.logger.debug(f"getActionList, checking  bundleId = {bundleId}")
                        tree = ET.parse(f"{indigoInstallPath}/Plugins/{plugin}/Contents/Server Plugin/Actions.xml")
                        actions = tree.getroot()
                        for action in actions:
                            if action.tag == "Action":
                                self.logger.debug(f"getActionList, Action attribs = {action.attrib}")
                                name = action.find('Name')
                                callBack = action.find('CallbackMethod')
                                if name is not None and callBack is not None:
                                    self.logger.debug(f"getActionList, Action id = {action.attrib['id']}, name = '{name.text}', callBackMethod = {callBack.text}")
                                    retList.append((action.attrib["id"], name.text))
```

**Comment by FlyingDiver** [comment]
Might be a bad upload. Will look again when I get a chance.

**Comment by FlyingDiver** [comment]
Fixed in 2022.0.3
